geoip: read source IP from EDNS0 subnet if provided (#5183)
miekg/coredns: plugin/geoip/README.md

// ...
if getLongitude := metadata.ValueFunc(ctx, "geoip/longitude"); getLongitude != nil {
    if longitude, err := strconv.ParseFloat(getLongitude(), 64); err == nil {
        // Do something useful with longitude.
    }
} else {
    // The metadata label geoip/longitude for some reason, was not set.
}
// ...
```

## Databases

The supported databases use city schema such as `City` and `Enterprise`. Other databases types with different schemas are not supported yet.

You can download a [free and public City database](https://dev.maxmind.com/geoip/geolite2-free-geolocation-data).

## Syntax

```text
geoip [DBFILE]
```

or

```text
geoip [DBFILE] {

>     edns-subnet-processing

Reviewer: ```suggestion
    [edns-subnet-processing]
```

Reviewer: Another _very_ minor thing: `edns-subnet-processing` is kind of a long option name.  e.g. I think just `edns-subnet` would be sufficient.

Author: `edns-subnet-processing` is taken from PowerDNS options, verbatim. We're not tied to pdns though :) I'm editing it shortly.

}
```

> By default, geoip plugin doesn't honor EDNS0 subnet extension, but it can be enabled using the extended syntax:

Reviewer: ```suggestion
```

Author: huh did that line make it so far?

Reviewer: I think that this is described sufficiently in the syntax section, and doesn't require more documentation.

* **DBFILE** the mmdb database file path.

> * `edns-subnet-processing` (optional): by default, geoip plugin read client IP and ignores [EDNS0 subnet](https://en.wikipedia.org/wiki/EDNS_Client_Subnet) extension. Using this option enables reading source IP from EDNS0 subnet when available. This help identifying closest source IP address as well as making GeoIP test processing easy: `dig +subnet=1.2.3.4 @dns-server.example.com www.geo-aware.com`.

Reviewer: ```suggestion
* `edns-subnet-processing`  Optional. For the geo IP address, use [EDNS0 subnet](https://en.wikipedia.org/wiki/EDNS_Client_Subnet) (if present) instead of the source IP of the DNS request. This helps with identifying the closest source IP address as well as making GeoIP testing easy: `dig +subnet=1.2.3.4 @dns-server.example.com www.geo-aware.com`.
```

It's not clear what you mean by "This helps with identifying the closest source IP address".

Author: @chrisohaver I'm theorycrafting here, but in theory, public recursors can forward original requestors' IP addresses in an EDNS0 subnet attachment. We just open the gates for such a behavior.

Otherwise, good changes.

## Examples

The following configuration configures the `City` database.

```txt
. {
    geoip /opt/geoip2/db/GeoLite2-City.mmdb {

>       edns-subnet-processing

Reviewer: The description of the example should reflect this. e.g. "The following configuration configures the `City` database, and looks up geo location based on edns0 subnet if present.

Author: Good point, thanks.

Author: I was reusing this conversation for edns-subnet processing benefits. Now a possible benefit for intermediary dns recursors is mentioned at "edns-subnet" config option.

* **DBFILE** the mmdb database file path.

By default, geoip plugin doesn't honor EDNS0 subnet extension, but it can be enabled using the extended syntax:

```
geoip [DBFILE] {
    edns-subnet-processing
}
```

> * **DBFILE**: see above

Reviewer 2: **Nit**

I find adding once again here DBFILE argument a bit redundant, given that is already defined above. I also changed the wording along with syntax and examples section... What do you think?

## Syntax
```txt
geoip [DBFILE] {
    edns-subnet-processing
}
```

* **DBFILE** the mmdb database file path.
* `edns-subnet-processing` By default, geoip plugin read client IP and ignores EDNS0 subnet extension, using this option enable reading source IP from [EDNS0 subnet](https://en.wikipedia.org/wiki/EDNS_Client_Subnet) when available. This help identifying closest source IP address as well as making GeoIP test processing easy: `dig +subnet=1.2.3.4 @dns-server.example.com www.geo-aware.com`.

## Examples

The following configuration configures the `City` database and enables EDNS subnet processing.

```txt
. {
    geoip /opt/geoip2/db/GeoLite2-City.mmdb {
        edns-subnet-processing
    }
    metadata # Note that metadata plugin must be enabled as well.
}
```

Author: I think it's a good idea, but I'd mention that `edns-subnet-processing` is optional, as well as the { } block. Otherwise, I'm happy with the rewording.

Reviewer 2: From my side, sounds good to be explicit with the optional nature of the argument.

Author: Please have a look, and open a conversation if something else is needed.

```

* **DBFILE** the mmdb database file path.

> * `edns-subnet`: Optional. Use [EDNS0 subnet](https://en.wikipedia.org/wiki/EDNS_Client_Subnet) (if present) for Geo IP instead of the source IP of the DNS request. This helps identifying the closest source IP address through intermediary DNS resolvers, and it also makes GeoIP testing easy: `dig +subnet=1.2.3.4 @dns-server.example.com www.geo-aware.com`.

Reviewer: I think "intermediary DNS resolvers" might be technically imperfect description here  - _but I think its fine as is_.  The specific situation that ends-subnet aims to assist is when any intermediate network "stuff" changes the source ip address of the request packet in transit, thus obfuscating the original client's IP.  More specifically, when the new IP address is not "near" the client.

IOW - it's not just "intermediate DNS resolvers" that can change the IP, other common situations can change the source IP, and ECS also solves those situations.

Author: I agree, it's a nice 80% solution :) I'm fine with that though.
